rasa_nlu_examples/featurizers/sparse/hashing_featurizer.py

```diff
@@ -0,0 +1,151 @@
+
+    The implementation used is Scikit-Learn's HashingVectorizer class, which is described in detail here:
+    https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html
+    """
+
+    @classmethod
+    def required_components(cls) -> List[Type[Component]]:
+        return [Tokenizer]
+
+    @classmethod
+    def required_packages(cls) -> List[Text]:
+        return ["sklearn"]
+
+    defaults = {
+        # The following parameters and defaults are similar as the ones used by
+        # scikit-learn (version 0.24.2). For detailed explanations on these parameters
+        # and their defaults, have a look at the scikit-learn docs:
+        # https://scikit-learn.org/0.24/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html
+        #
+        # the number of columns in the output vector
+        "n_features": 2048,  # int
+        # whether to use word or character n-grams
+        # 'char_wb' creates character n-grams inside word boundaries
+        # n-grams at the edges of words are padded with space.
+        "analyzer": "word",  # use 'char' or 'char_wb' for character
+        # remove accents during the preprocessing step
+        "strip_accents": None,  # {'ascii', 'unicode', None}
+        # list of stop words
+        "stop_words": None,  # string {'english'}, list, or None (default)
+        # if `True` convert all characters to lowercase
+        "lowercase": True,  # bool
+        # set range of ngrams to be extracted
+        "min_ngram": 1,
+        "max_ngram": 1,
+        # if `True`, all non zero counts are set to 1.
+        "binary": False,  # bool
+        # the norm used to normalize term vectors
+        "norm": "l2",  # 'l1', 'l2', or None
+        # when `True`, an alternating sign is added to the features as to approximately
+        # conserve the inner product in the hashed space even for small n_features.
+        "alternate_sign": True,  # bool
+    }
+
+    def __init__(self, component_config: Optional[Dict[Text, Any]] = None) -> None:
+        super().__init__(component_config)
+        self.hashing_vectorizer = HashingVectorizer(
+            n_features=self.component_config["n_features"],
+            analyzer=self.component_config["analyzer"],
+            token_pattern=r"(?u)\b\w+\b"
+            if self.component_config["analyzer"] == "word"
+            else None,
+            strip_accents=self.component_config["strip_accents"],
+            stop_words=self.component_config["stop_words"],
+            lowercase=self.component_config["lowercase"],
+            ngram_range=(
+                self.component_config["min_ngram"],
+                self.component_config["max_ngram"],
+            ),
+            binary=self.component_config["binary"],
+            norm=self.component_config["norm"],
+            alternate_sign=self.component_config["alternate_sign"],
+            dtype=np.float32,
```

Comment:
nit: we could use dictionary unpacking for this, e.g. `range = .... component_config.pop(<min/max>),  { **self.component_config, token_pattern = ..., ngram_range = range, dtype=np.float32}` (and maybe adapt the parameters of the component to be 1:1 like in the expected sklearn version (`ngram_range`), or does that help to make it work for several versions or something? 🤔 )

Reply:
This is mimicking the `CountVectorsFeaturizer` implementation [here](https://github.com/RasaHQ/rasa/blob/main/rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py#L738). In my understanding this is the best way to put it in YAML, since with e.g.
```
- name: HashingFeaturizer
  ngram_range: (1, 4)
```
the `(1, 4)` would be parsed as string and would have to be deconstructed into the min/max. Other option I could think of would be
```
- name: HashingFeaturizer
  ngram_range:
    - 1
    - 4
```
but that could be confusing, since it can be only two values and they need to be ordered. So being explicit about min/max seemed like the best option.

Extracting the range before would the dict unpacking work, however `self.component_config` also contains entries `name` and `alias` that are added by the base classes, so those would have to be filtered before the unpacking.

Reply:
- on the min/max: agree, that doesn't looks so nice in yaml and so we should keep the explicit params
- on the unpacking: we can filter the dict with the keys in the default config, pop the range which we know is encoded differently and then use unpacking on the remaining dict -> doesn't require us to know which keys might be used elsewhere

(btw, ideally, I'd like grab the default params from sklearn and build the default dict from that to always use the right parameters - which we should get to work in rasa 3 :D)